File: src/ch/jbead/Rect.java
/** jbead - http://www.brunoldsoftware.ch
    Copyright (C) 2001-2012  Damian Brunold

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

package ch.jbead;

import java.util.Iterator;

public class Rect implements Iterable<Point> {

    protected Point begin;
    protected Point end;

    public Rect(Point begin, Point end) {
        this.begin = begin;
        this.end = end;
    }

    @Override
    public Iterator<Point> iterator() {
        return new RectIterator(begin, end);
    }

    @Override
    public String toString() {
        return begin + "-" + end;
    }

    public int left() {
        return getBegin().getX();
    }

    public int right() {
        return getEnd().getX();
    }

    public int bottom() {
        return getBegin().getY();
    }

    public int top() {
        return getEnd().getY();
    }

    public int width() {
        return right() - left() + 1;
    }

    public int height() {
        return top() - bottom() + 1;
    }

    public Point getBegin() {
        return new Point(Math.min(begin.getX(), end.getX()), Math.min(begin.getY(), end.getY()));
    }

    public Point getEnd() {
        return new Point(Math.max(begin.getX(), end.getX()), Math.max(begin.getY(), end.getY()));
    }

    public boolean isSquare() {
        return Math.abs(end.getX() - begin.getX()) == Math.abs(end.getY() - begin.getY());
    }

    public boolean isColumn() {
        return begin.getX() == end.getX();
    }

    public boolean isRow() {
        return begin.getY() == end.getY();
    }

    public int size() {
        return width() * height();
    }

    public Rect scrolled(int scroll) {
        return new Rect(begin.scrolled(scroll), end.scrolled(scroll));
    }

}

File: csharp/JBead.Web/Core/Rect.cs
using System.Collections;

namespace JBead.Web.Core;

public class Rect : IEnumerable<Point>
{
    public static readonly Rect Empty = new(new Point(0, 0), new Point(-1, -1));

    protected Point begin;
    protected Point end;

    public Rect(Point begin, Point end)
    {
        this.begin = begin;
        this.end = end;
    }

    public IEnumerator<Point> GetEnumerator() => new RectIterator(begin, end);
    IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();

    public override string ToString() => $"{begin}-{end}";

    public int Left => Begin.X;
    public int Right => End.X;
    public int Bottom => Begin.Y;
    public int Top => End.Y;
    public int Width => Right - Left + 1;
    public int Height => Top - Bottom + 1;

    public Point Begin => new(Math.Min(begin.X, end.X), Math.Min(begin.Y, end.Y));
    public Point End => new(Math.Max(begin.X, end.X), Math.Max(begin.Y, end.Y));

    public bool IsSquare => Math.Abs(end.X - begin.X) == Math.Abs(end.Y - begin.Y);
    public bool IsColumn => begin.X == end.X;
    public bool IsRow => begin.Y == end.Y;
    public int Size => Width * Height;

    public Rect Scrolled(int scroll) => new(begin.Scrolled(scroll), end.Scrolled(scroll));
}
